Changes in module 'nf-core/ensemblvep/download'
--- modules/nf-core/ensemblvep/download/main.nf
+++ modules/nf-core/ensemblvep/download/main.nf
@@ -26,6 +26,7 @@
         --SPECIES $species \\
         --ASSEMBLY $assembly \\
         --CACHE_VERSION $cache_version \\
+        --NO_UPDATE \\
         $args
 
     cat <<-END_VERSIONS > versions.yml

************************************************************
