Changes in module 'nf-core/bedtools/intersect'
--- modules/nf-core/bedtools/intersect/main.nf
+++ modules/nf-core/bedtools/intersect/main.nf
@@ -8,33 +8,32 @@
         'biocontainers/bedtools:2.31.1--hf5e1c6e_0' }"
 
     input:
-    tuple val(meta), path(intervals1), path(intervals2)
-    tuple val(meta2), path(chrom_sizes)
+    tuple val(meta), path(bam), path(bai)
+    path(hk_bed)
 
     output:
-    tuple val(meta), path("*.${extension}"), emit: intersect
-    path  "versions.yml"                   , emit: versions
+    tuple val(meta), path("*_downsample_hk.bam"), emit: hk_bam   
+    path  "versions.yml"                        , emit: versions
 
     when:
-    task.ext.when == null || task.ext.when
+    //task.ext.when == null || task.ext.when
+    params.hk_bed
 
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
     //Extension of the output file. It is set by the user via "ext.suffix" in the config. Corresponds to the file format which depends on arguments (e. g., ".bed", ".bam", ".txt", etc.).
-    extension = task.ext.suffix ?: "${intervals1.extension}"
-    def sizes = chrom_sizes ? "-g ${chrom_sizes}" : ''
-    if ("$intervals1" == "${prefix}.${extension}" ||
-        "$intervals2" == "${prefix}.${extension}")
-        error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
+    //extension = task.ext.suffix ?: "${intervals1.extension}"
+    //def sizes = chrom_sizes ? "-g ${chrom_sizes}" : ''
+    //if ("$intervals1" == "${prefix}.${extension}" ||
+    //    "$intervals2" == "${prefix}.${extension}")
+    //    error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
     """
     bedtools \\
         intersect \\
-        -a $intervals1 \\
-        -b $intervals2 \\
-        $args \\
-        $sizes \\
-        > ${prefix}.${extension}
+        -a $bam \\
+        -b $hk_bed \\
+        > "${prefix}_downsample_hk.bam"
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
@@ -44,12 +43,12 @@
 
     stub:
     def prefix = task.ext.prefix ?: "${meta.id}"
-    extension = task.ext.suffix ?: "bed"
-    if ("$intervals1" == "${prefix}.${extension}" ||
-        "$intervals2" == "${prefix}.${extension}")
-        error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
+    //extension = task.ext.suffix ?: "bed"
+    //if ("$intervals1" == "${prefix}.${extension}" ||
+    //    "$intervals2" == "${prefix}.${extension}")
+    //    error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
     """
-    touch ${prefix}.${extension}
+    touch ${prefix}_downsample_hk.bam
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
