Changes in module 'nf-core/rseqc/junctionsaturation'
--- modules/nf-core/rseqc/junctionsaturation/main.nf
+++ modules/nf-core/rseqc/junctionsaturation/main.nf
@@ -8,7 +8,7 @@
         'biocontainers/rseqc:5.0.3--py39hf95cd2a_0' }"
 
     input:
-    tuple val(meta), path(bam)
+    tuple val(meta), path(bam), path(bai)
     path  bed
 
     output:

************************************************************
